File: aermet/core/geocode.py
"""
Geocoding module — resolves user-provided locations to lat/lon.

Supports:
  - Street address (via Nominatim OpenStreetMap API)
  - Direct lat/lon entry
  - ICAO airport code (bundled database)
"""

import csv
import json
import math
import os

from .http import get as http_get

# Bundled airport database path (relative to this file)
_DATA_DIR = os.path.join(os.path.dirname(__file__), '..', 'data')


def geocode_address(address):
    """
    Geocode a street address using Nominatim (OpenStreetMap).
    Returns dict with lat, lon, display_name or raises ValueError.
    """
    url = 'https://nominatim.openstreetmap.org/search'
    params = {
        'q': address,
        'format': 'json',
        'limit': 1,
        'addressdetails': 1,
    }
    headers = {'User-Agent': 'AERMET-Automation-Tool/1.0'}
    resp = http_get(url, params=params, headers=headers, timeout=15)
    resp.raise_for_status()
    results = resp.json()
    if not results:
        raise ValueError(f'No results found for address: {address}')

    r = results[0]
    country_code = r.get('address', {}).get('country_code', '')
    return {
        'lat': float(r['lat']),
        'lon': float(r['lon']),
        'display_name': r.get('display_name', ''),
        'country_code': country_code.upper(),
        'is_us': country_code.lower() == 'us',
    }


def geocode_airport(icao_code):
    """
    Look up an ICAO airport code from the bundled database.
    Returns dict with lat, lon, name, country_code, is_us.
    """
    icao_code = icao_code.strip().upper()
    db_path = os.path.join(_DATA_DIR, 'airports.csv')

    if not os.path.exists(db_path):
        # Fall back to Nominatim
        return geocode_address(f'{icao_code} airport')

    with open(db_path, 'r', encoding='utf-8') as f:
        reader = csv.DictReader(f)
        for row in reader:
            if row.get('icao', '').strip().upper() == icao_code:
                cc = row.get('country', '').strip().upper()
                return {
                    'lat': float(row['lat']),
                    'lon': float(row['lon']),
                    'name': row.get('name', icao_code),
                    'country_code': cc,
                    'is_us': cc == 'US',
                }

    raise ValueError(f'Airport code {icao_code} not found in database')


def geocode_latlon(lat, lon):
    """
    Validate and return a direct lat/lon entry.
    Determines country via simple bounding box (US check).
    """
    lat = float(lat)
    lon = float(lon)
    if lat < -90 or lat > 90:
        raise ValueError(f'Invalid latitude: {lat}')
    if lon < -180 or lon > 180:
        raise ValueError(f'Invalid longitude: {lon}')

    is_us = _is_conus(lat, lon)
    return {
        'lat': lat,
        'lon': lon,
        'display_name': f'{lat:.4f}, {lon:.4f}',
        'country_code': 'US' if is_us else '',
        'is_us': is_us,
    }


def _is_conus(lat, lon):
    """Simple bounding-box check for continental US."""
    return 24.0 <= lat <= 50.0 and -125.0 <= lon <= -66.0


def reverse_geocode(lat, lon):
    """Reverse geocode to determine country."""
    url = 'https://nominatim.openstreetmap.org/reverse'
    params = {'lat': lat, 'lon': lon, 'format': 'json', 'zoom': 3}
    headers = {'User-Agent': 'AERMET-Automation-Tool/1.0'}
    try:
        resp = http_get(url, params=params, headers=headers, timeout=15)
        resp.raise_for_status()
        data = resp.json()
        cc = data.get('address', {}).get('country_code', '').upper()
        return {'country_code': cc, 'is_us': cc == 'US'}
    except Exception:
        return {'country_code': '', 'is_us': _is_conus(lat, lon)}


def haversine_km(lat1, lon1, lat2, lon2):
    """Distance in km between two lat/lon points."""
    R = 6371.0
    rlat1, rlon1 = math.radians(lat1), math.radians(lon1)
    rlat2, rlon2 = math.radians(lat2), math.radians(lon2)
    dlat = rlat2 - rlat1
    dlon = rlon2 - rlon1
    a = math.sin(dlat / 2) ** 2 + math.cos(rlat1) * math.cos(rlat2) * math.sin(dlon / 2) ** 2
    return R * 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))

File: aermet-tool/src/geocode.rs
/// Geocoding — resolves user-provided locations to lat/lon.
///
/// Supports: street address (Nominatim), ICAO airport code (embedded CSV), direct lat/lon.
use crate::http_client::client;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GeoResult {
    pub lat: f64,
    pub lon: f64,
    #[serde(default)]
    pub display_name: String,
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub country_code: String,
    pub is_us: bool,
}

/// Geocode a street address via Nominatim (OpenStreetMap).
pub async fn geocode_address(address: &str) -> Result<GeoResult, String> {
    let resp = client()
        .get("https://nominatim.openstreetmap.org/search")
        .query(&[
            ("q", address),
            ("format", "json"),
            ("limit", "1"),
            ("addressdetails", "1"),
        ])
        .timeout(std::time::Duration::from_secs(15))
        .send()
        .await
        .map_err(|e| format!("Geocode request failed: {e}"))?;

    let results: Vec<serde_json::Value> = resp
        .json()
        .await
        .map_err(|e| format!("Geocode parse failed: {e}"))?;

    let r = results
        .first()
        .ok_or_else(|| format!("No results found for address: {address}"))?;

    let lat: f64 = r["lat"]
        .as_str()
        .unwrap_or("0")
        .parse()
        .unwrap_or(0.0);
    let lon: f64 = r["lon"]
        .as_str()
        .unwrap_or("0")
        .parse()
        .unwrap_or(0.0);
    let display_name = r["display_name"].as_str().unwrap_or("").to_string();
    let country_code = r
        .get("address")
        .and_then(|a| a.get("country_code"))
        .and_then(|c| c.as_str())
        .unwrap_or("")
        .to_uppercase();

    Ok(GeoResult {
        lat,
        lon,
        display_name,
        name: String::new(),
        country_code: country_code.clone(),
        is_us: country_code == "US",
    })
}

/// Look up an ICAO airport code from the embedded CSV database.
pub async fn geocode_airport(icao_code: &str) -> Result<GeoResult, String> {
    let code = icao_code.trim().to_uppercase();
    let csv_data = include_str!("../data/airports.csv");

    let mut rdr = csv::Reader::from_reader(csv_data.as_bytes());
    for record in rdr.records() {
        let record = record.map_err(|e| format!("CSV parse error: {e}"))?;
        let icao = record.get(5).unwrap_or("").trim().to_uppercase(); // icao column
        if icao == code {
            let lat: f64 = record.get(6).unwrap_or("0").trim().parse().unwrap_or(0.0);
            let lon: f64 = record.get(7).unwrap_or("0").trim().parse().unwrap_or(0.0);
            let name = record.get(1).unwrap_or("").trim().to_string();
            let cc = record.get(3).unwrap_or("").trim().to_uppercase();
            return Ok(GeoResult {
                lat,
                lon,
                display_name: name.clone(),
                name,
                country_code: cc.clone(),
                is_us: cc == "US",
            });
        }
    }

    // Fall back to Nominatim
    geocode_address(&format!("{code} airport")).await
}

/// Validate and return a direct lat/lon entry.
pub fn geocode_latlon(lat: f64, lon: f64) -> Result<GeoResult, String> {
    if !(-90.0..=90.0).contains(&lat) {
        return Err(format!("Invalid latitude: {lat}"));
    }
    if !(-180.0..=180.0).contains(&lon) {
        return Err(format!("Invalid longitude: {lon}"));
    }

    let is_us = is_conus(lat, lon);
    Ok(GeoResult {
        lat,
        lon,
        display_name: format!("{lat:.4}, {lon:.4}"),
        name: String::new(),
        country_code: if is_us { "US".to_string() } else { String::new() },
        is_us,
    })
}

/// Simple bounding-box check for continental US.
fn is_conus(lat: f64, lon: f64) -> bool {
    (24.0..=50.0).contains(&lat) && (-125.0..=-66.0).contains(&lon)
}

/// Haversine distance in km between two lat/lon points.
pub fn haversine_km(lat1: f64, lon1: f64, lat2: f64, lon2: f64) -> f64 {
    const R: f64 = 6371.0;
    let rlat1 = lat1.to_radians();
    let rlon1 = lon1.to_radians();
    let rlat2 = lat2.to_radians();
    let rlon2 = lon2.to_radians();
    let dlat = rlat2 - rlat1;
    let dlon = rlon2 - rlon1;
    let a = (dlat / 2.0).sin().powi(2) + rlat1.cos() * rlat2.cos() * (dlon / 2.0).sin().powi(2);
    R * 2.0 * a.sqrt().atan2((1.0 - a).sqrt())
}
